fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -231,8 +252,51 @@ private boolean shouldClearExistingIndexEntries(@Nonnull IndexState indexState,
         }
     }
 
+    @Nonnull
+    public IndexingBase setFallbackMode() {
+        fallbackMode = true;
+        return this;
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> setIndexingTypeOrThrow() {
+        byte[] stamp = getTypeStamp();
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "setIndexingTypeOrThrow",
+                LogMessageKeys.LOCAL_VERSION, stamp);
+
+        return buildCommitRetryAsync( (store, ignore) -> {
+            Transaction transaction = store.getContext().ensureActive();
+            byte[] stampKey = indexBuildTypeSubspace(store, common.getIndex()).getKey();
+            return transaction.get(stampKey)
+                            .thenApply( bytes -> {
+                                if (bytes == null || fallbackMode) {
+                                    transaction.set(stampKey, stamp);
+                                } else if (! matchingTypeStamp(bytes)) {
+                                    if (LOGGER.isWarnEnabled()) {
+                                        LOGGER.warn(KeyValueLogMessage.of("Index was partly built by another method",
+                                                LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
+                                                LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
+                                                LogMessageKeys.INDEXER_ID, common.getUuid(),
+                                                LogMessageKeys.EXPECTED_TYPE, stamp,
+                                                LogMessageKeys.ACTUAL_TYPE, bytes));
+                                    }
+                                    throw new RecordCoreException("This index was partly built by another method");
+                                }
+                                return null;
+                            }); },
+                false,
+                additionalLogMessageKeyValues);
+    }
+
+    @Nonnull
+    abstract byte[] getTypeStamp();
+
+    @Nonnull
+    abstract boolean matchingTypeStamp(byte[] stamp);
+
     abstract CompletableFuture<Void> buildIndexByEntityAsync();
```

Comment:
What is an "entity" in this context? I guess it's not new to this branch, but it seems like a weird word to use.

Reply:
At this point, "entity" is record or index. How would you rename it?